libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/location/CarLocationProvider.kt

```diff
@@ -0,0 +1,77 @@
+package com.mapbox.androidauto.car.location
+
+import android.location.Location
+import com.mapbox.maps.plugin.locationcomponent.LocationConsumer
+import com.mapbox.maps.plugin.locationcomponent.LocationProvider
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationApp
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationObserver
+import com.mapbox.navigation.core.trip.session.LocationMatcherResult
+import com.mapbox.navigation.core.trip.session.LocationObserver
+import com.mapbox.navigation.ui.maps.location.NavigationLocationProvider
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.firstOrNull
+
+/**
+ * Automatically attaches and detaches from [MapboxNavigationApp] to provide map matched locations.
+ */
+class CarLocationProvider private constructor() : MapboxNavigationObserver, LocationProvider {
+
+    private val navigationLocationProvider = NavigationLocationProvider()
+    private val mutableLocation = MutableSharedFlow<Location?>(
+        replay = 1,
+        onBufferOverflow = BufferOverflow.DROP_OLDEST
+    )
+
+    private val locationObserver = object : LocationObserver {
+        override fun onNewLocationMatcherResult(locationMatcherResult: LocationMatcherResult) {
+            navigationLocationProvider.changePosition(
+                locationMatcherResult.enhancedLocation,
+                locationMatcherResult.keyPoints,
+            )
+            mutableLocation.tryEmit(locationMatcherResult.enhancedLocation)
+        }
+
+        override fun onNewRawLocation(rawLocation: Location) {
+            // no op
+        }
+    }
+
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        mapboxNavigation.registerLocationObserver(locationObserver)
+    }
+
+    override fun onDetached(mapboxNavigation: MapboxNavigation) {
+        mapboxNavigation.unregisterLocationObserver(locationObserver)
+    }
+
+    override fun registerLocationConsumer(locationConsumer: LocationConsumer) {
+        navigationLocationProvider.registerLocationConsumer(locationConsumer)
+    }
+
+    override fun unRegisterLocationConsumer(locationConsumer: LocationConsumer) {
+        navigationLocationProvider.unRegisterLocationConsumer(locationConsumer)
+    }
+
+    /**
+     * Immediately access the last location received.
+     */
+    val lastLocation: Location?
+        get() = navigationLocationProvider.lastLocation
+
+    /**
+     * Wait until a non-null location is received. Improves results when the app is starting.
+     */
+    suspend fun validLocation(): Location? = mutableLocation.firstOrNull()
```

Comment:
Can be non-null, because `SharedFlow` never completes. 
```suggestion
    suspend fun validLocation(): Location = mutableLocation.first()
```